File: Solutions/Python/longest_substring_without_repeating_characters.py
"""
    [MEDIUM]
    3. Longest Substring Without Repeating Characters

    Concepts:
    - sliding window
    - hash table

    Stats:
        Runtime | 15 ms     [Beats 81.90%]
        Memory  | 18.04 MB  [Beats 6.50%]
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest = 0
        start = 0
        seen = set()

        for end in range(len(s)):
            while s[end] in seen:
                seen.remove(s[start])
                start += 1

            seen.add(s[end])

            longest = max(longest, end - start + 1)

        return longest

File: Solutions/Rust/longest_substring_without_repeating_characters.rs
/*
    [MEDIUM]
    3. Longest Substring Without Repeating Characters

    Concepts:
    - sliding window
    - hash table

    Stats:
        Runtime | 1 ms      [Beats 60.29%]
        Memory  | 2.32 MB   [Beats 32.19%]
*/
use std::collections::HashSet;
use std::cmp::max;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut longest = 0;
        let mut start = 0;
        let mut seen = HashSet::new();
        let chars = s.chars().collect::<Vec<char>>();

        for end in 0..chars.len() {
            while seen.contains(&chars[end]) {
                seen.remove(&chars[start]);
                start += 1
            }

            seen.insert(chars[end]);
            longest = max(longest, end - start + 1);
        }

        return longest as i32;
    }
}